### PR TITLE
Adds HttpClient, HttpException, and HttpResonse classes

### PR DESCRIPTION
**Pull request by jonahwilliams**

### CODE REVIEW

**Comment by jonahwilliams** [comment]
cc @leonsenft

File: lib/web_http.dart [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+  const HttpException._(this.statusText, this.statusCode);
+
+  /// The status code from the http request.
+  final int statusCode;
+
+  /// The status text from the http request.
+  final String statusText;
+
+  @override
+  String toString() => statusText;
```
Comment (leonsenft) [inline lib/web_http.dart]: Does the status text contain the status code? If not, should we include the status code in toString()?

File: pubspec.yaml [hunk 2]
```diff
@@ -0,0 +1,8 @@
+name: web_http
+description: A web http client focused on speed and correctness.
+version: 0.0.1
+#homepage: https://www.example.com
```
Comment (leonsenft) [inline pubspec.yaml]: Point homepage to repository?

File: pubspec.yaml [hunk 3]
```diff
@@ -0,0 +1,8 @@
...
+author: Jonah Williams <jonahwilliams@google.com>
+
+environment:
+  sdk: '>=1.20.1 <2.0.0'
```
Comment (leonsenft) [inline pubspec.yaml]: Why not start on 2.0.0-dev?

File: README.md [hunk 4]
```diff
@@ -0,0 +1,3 @@
+# web_http
+
+A web http client focused on speed and correctness.
```
Comment (leonsenft) [inline README.md]: nit: HTTP

**Comment by leonsenft** [comment]
Have you considered a testing story for this yet? Do you think it's worth shipping a test implementation of the client?

File: pubspec.yaml (hunk 3 above)
Comment (jonahwilliams) [reply pubspec.yaml]: Does that work, just 'sdk: 2.0.0-dev' ?

**Comment by jonahwilliams** [comment]
Yes, I think this absolutely needs to included a testing implementation.  However, in terms of testing the client itself, I really have no idea what the best approach is.  Configuring a local server seems to be the only way to do that.

Will update with testing implementation

File: lib/web_http.dart [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+    int timeout,
+    bool withCredentials,
+  }) {
+    return request(url, 'POST', headers: headers, timeout: timeout)
+        .map((HttpResponse response) => response.body);
+  }
+
+  /// Sends an http request to a specified domain.
+  ///
+  /// [url] is the String encoded url.
```
Comment (leonsenft) [inline lib/web_http.dart]: nit: Your args comments aren't formatted consistently here in comparison to the previous two methods.

File: lib/web_http.dart [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+  /// [body] is the (optional) request body.
+  /// [headers] are the (optional) response headers.
+  /// [timeout] is an (optional) time limit in milliseconds before aborting
+  /// the request automatically.
+  /// [withCredentials] whether cross-site requests should use cookie or
+  /// header credentials.
+  Stream<HttpResponse> request(
+    String url,
+    String method, {
+    String responseType,
```
Comment (leonsenft) [inline lib/web_http.dart]: Seem to be missing a description for this in the doc comment.

File: lib/web_http.dart [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+      final exception = new HttpException._(
+        request.statusText,
+        request.status,
+      );
+      controller.addError(exception);
+      controller.close();
+    });
+    request.onLoadEnd.listen((ProgressEvent event) {
+      if (controller.isClosed) return;
+      if (request.status == 200) {
```
Comment (leonsenft) [inline lib/web_http.dart]: Aren't there other non-error response codes too?

Comment (leonsenft) [reply lib/web_http.dart]: "201 Created" for POST or PUT for example.

File: pubspec.yaml (hunk 3 above)
Comment (leonsenft) [reply pubspec.yaml]: I think you want ^2.0.0-dev to accept all dev versions.

**Comment by leonsenft** [comment]
In terms of testing, do you think providing a test implementation of the client itself is fine, or would it be better to have the client accept an HttpRequest factory function? That way even the test code runs through all of the same request field setup and callback registration.

File: lib/web_http.dart [hunk 8]
```diff
@@ -40,16 +72,16 @@ class HttpClient {
...
-  /// [method] is an http verb such as 'GET' or 'POST'.
-  /// [body] is the (optional) request body.
-  /// [headers] are the (optional) response headers.
-  /// [timeout] is an (optional) time limit in milliseconds before aborting
-  /// the request automatically.
-  /// [withCredentials] whether cross-site requests should use cookie or
-  /// header credentials.
+  ///  * [url] is the string encoded url.
+  ///  * [method] is an http verb such as 'GET' or 'POST'.
+  ///  * [body] is the (optional) request body.
```
Comment (leonsenft) [inline lib/web_http.dart]: Still missing documentation for [responseType].

File: lib/web_http.dart [hunk 9]
```diff
@@ -91,13 +123,13 @@ class HttpClient {
     });
     request.onLoadEnd.listen((ProgressEvent event) {
       if (controller.isClosed) return;
-      if (request.status == 200) {
+      if (request.status >= 200 && request.status < 300) {
```
Comment (leonsenft) [inline lib/web_http.dart]: I'm not an expert on HTTP codes, but aren't 3xx codes considered successful too?

File: lib/web_http.dart [hunk 10]
```diff
@@ -120,27 +152,27 @@ class HttpClient {
...
-/// An http response containing the text response body and headers.
+/// An HTTP response containing the text response body and headers.
 class HttpResponse {
   const HttpResponse._(this.body, this.headers);
 
   /// The body of the response.
   ///
-  /// Depending on the respond type requested, might be a String, Document,
+  /// Depending on the respond type requested, might be a [String], [Document],
   /// or a buffer.
```
Comment (leonsenft) [inline lib/web_http.dart]: What's the actual type of a buffer?

File: lib/testing.dart [hunk 11]
```diff
@@ -0,0 +1,133 @@
+import 'dart:async';
+
+import 'package:web_http/web_http.dart';
+
+/// A signature for callback which is invoked whenever the [TestHttpClient]
+/// recieves a request.
+typedef FutureOr<TestHttpResponse> TestClientCallback(String url, String method,
```
Comment (leonsenft) [inline lib/testing.dart:7]: Consider using the new typedef syntax? 

```dart
typedef RequestHandler = FutureOr<TestHttpResonse> Function(
  String url,
  String method, {
  String responseType,
  String body,
  Map<String, String> headers,
  int timeout,
  bool withCredentials,
});
```

File: lib/testing.dart [hunk 12]
```diff
@@ -0,0 +1,133 @@
...
+  /// Non 2XX status code will cause the [TestHttpClient] to return an
+  /// [HttpException].
+  final int statusCode;
+}
+
+/// Helper function which throws an [UnsupportedError] if the [TestHttpClient] or
+/// [TestHttpResponse] classes are created in production mode.
+void _throwOnProductionMode() {
+  bool inProductionMode = true;
+  assert(() {
```
Comment (leonsenft) [inline lib/testing.dart]: You can shorten this to

```dart
assert(inProductionMode = false);
```

and get the same effect.

File: lib/testing.dart [hunk 13]
```diff
@@ -0,0 +1,133 @@
...
+    String body,
+    Map<String, String> headers,
+    int timeout,
+    bool withCredentials});
+
+/// A testing implementation of [HttpClient].
+/// 
+/// The exact responses returned can be controlled with the provided 
+/// [callback].
+class TestHttpClient implements HttpClient {
```
Comment (leonsenft) [inline lib/testing.dart]: Would it be a good idea to extend rather than implement `HttpClient` so you don't need to re-implement `get` and `post`?

Comment (leonsenft) [reply lib/testing.dart]: Not that they're much work to re-implement, but is there an advantage to implementing over extending?

File: lib/testing.dart [hunk 14]
```diff
@@ -0,0 +1,133 @@
...
+
+  @override
+  Stream<HttpResponse> request(String url, String method,
+      {String responseType,
+      String body,
+      Map<String, String> headers,
+      int timeout,
+      bool withCredentials}) {
+    StreamController controller;
+    controller = new StreamController(
```
Comment (leonsenft) [inline lib/testing.dart:34]: nit: collapse declaration and definition into same line:

```dart
final controller = new StreamController(
```

File: lib/testing.dart [hunk 15]
```diff
@@ -0,0 +1,133 @@
...
+              controller.add(response);
+              controller.close();
+            } else {
+              controller.addError(
+                  new HttpException('test exception', response.statusCode));
+              controller.close();
+            }
+          } else {
+            (response as Future<TestHttpResponse>).then((response) {
+              if (response.statusCode >= 200 && response.statusCode < 300) {
```
Comment (leonsenft) [inline lib/testing.dart]: Why not make this block a private method or closure to handle the response and avoid duplicating the implementation?

File: lib/web_http.dart (hunk 10 above)
Comment (jonahwilliams) [reply lib/web_http.dart]: Added, ByteBuffer

File: lib/testing.dart (hunk 11 above)
Comment (jonahwilliams) [reply lib/testing.dart:7]: That unfortunately doesn't allow naming url and method.  I could also make everything named, but I want to keep it synced with the request method.

File: lib/testing.dart (hunk 12 above)
Comment (jonahwilliams) [reply lib/testing.dart]: Done

File: lib/testing.dart (hunk 14 above)
Comment (jonahwilliams) [reply lib/testing.dart:34]: This is a trick to allow me to use `controller` in the `onListen` callback - otherwise, it hasn't been declared yet.

File: lib/testing.dart (hunk 15 above)
Comment (jonahwilliams) [reply lib/testing.dart]: Fixed with a Future.value

File: lib/testing.dart (hunk 13 above)
Comment (jonahwilliams) [reply lib/testing.dart]: Yeah, this seems like a brainfart on my part - fixed!

**Comment by jonahwilliams** [comment]
I have tried and failed to add tests, perhaps I will configure build_runner on v2
```
[Error from Dart2JS on web_http|test/web_http_test.dart.browser_test.dart]:
test/packages/term_glyph/src/generated.dart:136:5:
Expected ';' before this.
    _verticalLineQuadrupleDashBold = "┋";
    ^
```

File: lib/testing.dart (hunk 11 above)
Comment (leonsenft) [reply lib/testing.dart:7]: Oh really? That almost seems like a bug. :confused:

File: lib/testing.dart (hunk 14 above)
Comment (leonsenft) [reply lib/testing.dart:34]: Oh right!
